### ISSUE TITLE
Exports unclear

### ISSUE THREAD

**Issue by fsimonis** [precice/precice#1306]
**Describe your setup**

preCICE Version: 2.4.0

**Describe the problem**

preCICE allows exporting meshes of participants using the [configuration file](https://precice.org/configuration-export.html#structuring-exports).

Relevant options are:

| Option | Type | File Suffix | Comment |
| --- | --- | --- | --- |
| `every-n-time-windows` | Integer | `.dtN` | -1 means no exports |
| `every-iteration` | Boolean | `.itN` | |

This opens some questions:
* How can I export every time-step when sub-cycling?
* Does `every-n-time-windows` also influence when iterations are exported?
* Why are iterations globally counted? The preCICE logs report both time-window and iteration within the time-window, not the global iteration count.
* What does `every-n-time-windows=0` mean?
* When using `every-n-time-windows=3`, will the first and the 4th time-window be exported or the 3rd and the 6th?
* Also how can I omit/select meshes to export?

**Additional context**

#1300

**Comment by fsimonis** [comment]
Conclusions:
* The naming is tricky as iterations, dt and even Dt can change in various ways. We would need t instead. Best to only change this once it is necessary.
* Add filters for meshes `<export:vtu mesh="MeshOne" />`. Export everything by default. Filtering data isn't worth it. If we want to reduce the size of the individual files, then switching to binary representation will be a better way forward.
* Wait with the implementation of `until="X"` and `after="X"`, as it is not sure what unit we should use there. Iteration count may vary between time-windows, time-step count can vary between iterations, time could work, but again will not be unique as multiple iterations may use various time step sizes and produce different output. 

@Alphaoo1 the mesh tag should already solve your space problems.

**Comment by uekerman** [comment]
Sounds good in principle.

> Export everything by default

Should this include all iterations? This could be too much.
Does this mean to remove the option `every-n-time-windows`?

**Comment by fsimonis** [comment]
"everything" was meant only regarding meshes.

So use defaults as usual but export all meshes by default unless there is a specific mesh requested.

**Comment by uekerman** [comment]
Got it. I am sure this would be a nice and simple addition, but it does not solve this issue, correct?
(not saying we need to solve the issue now)

**Comment by uekerman** [comment]
There is a large overlap between this issue and #1655. I am replying here for both.

Let's simplify our lives:

- Only export data at the end of a time window. 
- Only export data once converged in an implicit coupling scheme. Thus, drop `every-iteration`.

Both are high effort low gain in my opinion. Only downside is that we cannot fully replay every waveform scenario with ASTE, but output at every time step can always be added later if ever needed.

> What does every-n-time-windows=0 mean?

Should lead to an error. We could also remove the `-1` option. I don't see a use case for this either.

> When using every-n-time-windows=3, will the first and the 4th time-window be exported or the 3rd and the 6th?

3rd and 6th. Have not checked, but I think this is also what other tools are doing. At least, that is what I would expect.

> Add filters for meshes <export:vtu mesh="MeshOne" />. Export all meshes by default.

:+1:

Could somebody else please take over the implementation? Should not too complicated.

**Comment by davidscn** [comment]
>     * Only export data at the end of a time window.
> 
>     * Only export data once converged in an implicit coupling scheme. Thus, drop `every-iteration`.
> 
> 
> Both are high effort low gain in my opinion. Only downside is that we cannot fully replay every waveform scenario with ASTE, but output at every time step can always be added later if ever needed.
> 

I understand that the every iteration option becomes incredible complex when combining it with every iteration. However, having the every iteration option there is still useful for adapter development when dealing with implicit coupling. There are also a few preCICE checks, which point to repeatedly written data and such things, which can easily be investigated using the every-iteration option.

I don't fully agree with the 'high-effort' argument to remove the feature, as the underlying functionality will remain. However, applying naming conventions for all combinations will become tricky. I would maybe consider the `every-iteration` as a separate case, with the only purpose of debugging. This way, we could just export files and name the files according to the iteration count or similar. The naming scheme would then be independent of all other export-naming logic.

**Comment by fsimonis** [comment]
An interesting note here is that the support for mixed time-window sizes #1713 makes the time window naming convention tricky to understand.

As an example:
* Given participants A, B, and C
* A is coupled to B with a time-window-size of 2
* B is coupled to C with a time-window-size of 3
* we use max-time 6

Then B has mixed time-window sizes, and we end up with exports at the following times:

Participant | t=2 | t=3 | t=4 | t=6 
---|---|---|---|---
A | tw1 |     | tw2 | tw3
B | tw1 | tw2 | tw3 | tw4
C |     | tw1 |     | tw2

**Comment by uekerman** [comment]
> An interesting note here is that the support for mixed time-window sizes https://github.com/precice/precice/issues/1713 makes the time window naming convention tricky to understand.

True. But also a rare case IMO. Remember that exporting is mainly a feature for debugging, not for production runs. There, users typically use the exporting of their solvers.

Let's only solve such problems once we need to.
